### PR TITLE
Removes Advanced Reagent Scanner from Service & Medical Cyborgs, Transfers Functionality to Innate Reagent Vision.

### PR DESCRIPTION
**Pull request by Fordoxia**
## What Does This PR Do
Removes the Advanced Reagent Scanner from the Service and Medical cyborg modules.

Reagent vision now shows % composition of chemicals in addition to absolute unit values (e.g: 10u water (50%), 10u cyanide (50%)).

Adds **Advanced Reagent Vision** to the game. It functions like normal reagent vision, but adds the functionality of the advanced reagent scanner.

Cyborgs with reagent vision have been upgraded to have advanced reagent vision.

Ghosts have been upgraded to have advanced reagent vision.

The debug HUD glasses has been given advanced reagent vision (it previously didn't even have normal reagent vision :/). The description has been updated to actually say everything it has.

NPF: Makes a bunch of 1/0 arguments into TRUE/FALSE arguments.

## Why It's Good For The Game
The medical borg's inventory is fairly big, anything to reduce it down is useful. Service borg is along for the ride.

Devs and admins fiddling with chemicals rejoice!

## Testing
Be medical borg. Didn't have scanner. Stared at various beakers, cigarettes, food items. Saw what I would have seen if I did have a scanner. All was good.

Tested regular science goggles, saw % amounts in addition to absolute units. All was good.

## Changelog
:cl:
add: Adds advanced reagent vision.
del: Medical and Service borgs don't have an Adv. Reagent scanner no more.
tweak: Medical and Service borgs now have advanced reagent vision. Debug HUD glasses have it too.
/:cl:

### CODE REVIEW

**Comment by FunnyMan3595** [comment]
> can see the exact chemical contents of anything by simply examining them

Are you **sure** about that?
image

**Comment by Fordoxia** [comment]
> > can see the exact chemical contents of anything by simply examining them
> 
> Are you **sure** about that?

Fair, but neither borg type has much business looking at pills anyway, considering they cannot handle them.

**Comment by Stoniest** [comment]
To be honest, I don't believe Borgs need to know the exact units in something, just the different reagents inside, I believe this to be a good change, less stuff in an already full toolbar is nothing but positive in that aspect, while they do lose the ability to know  the exact amount of units, that isn't too important for a borg, so it is a little down side, it is a pretty ignorable down side.

**Comment by Oyu07** [comment]
Just want to confirm, can the medical borg see the contents of a beaker with this pr

**Comment by Fordoxia** [comment]
> Just want to confirm, can the medical borg see the contents of a beaker with this pr
Yes 
image

**Comment by Fordoxia** [comment]
> > can see the exact chemical contents of anything by simply examining them
> 
> Are you **sure** about that?

As it turns out, the advanced reagent scanner only gives %amounts for pills, cigs, and food items too! There's ACTUALLY no loss of functionality vs just having reagent vision.

**Comment by FunnyMan3595** [comment]
> > > can see the exact chemical contents of anything by simply examining them
> > 
> > 
> > Are you **sure** about that?
> 
> As it turns out, the advanced reagent scanner only gives %amounts for pills, cigs, and food items too! There's ACTUALLY no loss of functionality vs just having reagent vision.

Reagent vision doesn't even show that; the output in my image was from using the advanced reagent scanner after examining each.

**Comment by FunnyMan3595** [comment]
I'm really not fond of losing the ability for service borgs to see what's in food and medborgs to see what's in pills/patches, so I'm against removing this until/unless that gets addressed.

**Comment by Fordoxia** [comment]
> I'm really not fond of losing the ability for service borgs to see what's in food and medborgs to see what's in pills/patches, so I'm against removing this until/unless that gets addressed.

I'll see if I can make reagent vision do the scanner stuff.

File: code/game/atoms.dm [hunk 1]
```diff
@@ -364,24 +364,54 @@
...
-						. += "<span class='notice'>[reagents.total_volume] units of various reagents.</span>"
-			else
-				. += "<span class='notice'>Nothing.</span>"
-		else if(container_type & AMOUNT_VISIBLE)
-			if(reagents.total_volume)
-				. += "<span class='notice'>It has [reagents.total_volume] unit\s left.</span>"
+			for(var/I in reagents.reagent_list)
+				var/datum/reagent/R = I
+				if(R.id != "blood")
+					. += "<span class='notice'>[R.volume] units of [R] ([R.volume / one_percent]%)</span>"
```
Comment (FunnyMan3595) [inline code/game/atoms.dm:380]:
```suggestion
					. += "<span class='notice'>[R.volume] units of [R] ([round(R.volume / one_percent, 0.1)]%)</span>"
```
And similar elsewhere.

**Comment by FunnyMan3595** [comment]
Design LGTM now.

File: code/game/atoms.dm (hunk 1 above)
Comment (Fordoxia) [reply code/game/atoms.dm]: Done.

File: code/datums/outfits/outfit_debug.dm [hunk 2]
```diff
@@ -95,10 +95,11 @@
 
 /obj/item/clothing/glasses/hud/debug
 	name = "AVD-CNED glasses"
-	desc = "Medical, security and diagnostic hud. Alt click to toggle xray."
+	desc = "Dagnostic, Hydroponic, Medical, Security, and Skills HUD. Built-in advanced reagent scanner. Alt click to toggle X-ray vision."
```
Comment (Henri215) [inline code/datums/outfits/outfit_debug.dm:98]:
```suggestion
	desc = "Diagnostic, Hydroponic, Medical, Security, and Skills HUD. Built-in advanced reagent scanner. Alt click to toggle X-ray vision."
```

File: code/game/atoms.dm [hunk 3]
```diff
@@ -390,24 +390,54 @@
...
 /atom/proc/build_reagent_description(mob/user)
 	. = list()
-	if(reagents)
+	if(!reagents)
+		return
+	var/one_percent = reagents.total_volume / 100
+	var/blood_type = ""
+	if(user.advanced_reagent_vision())	// You can see absolute unit concentrations in transparent containers and % concentrations in opaque containers. You can also see blood types.
+		. += "<span class='notice'>It contains:</span>"
+		if(!reagents.reagent_list.len)
```
Comment (Henri215) [inline code/game/atoms.dm:399]: image

Any reason why you changed it making it use .len?

Comment (Fordoxia) [reply code/game/atoms.dm]: That got missed during a merge deconflict, I've changed it back now.

File: code/game/atoms.dm [hunk 4]
```diff
@@ -390,24 +390,54 @@
...
+			. += "<span class='notice'>[reagents.total_volume] units of various reagents.</span>"
+			return
+		. += "<span class='notice'>Nothing.</span>"
+		return
+
+	if(container_type & AMOUNT_VISIBLE)
+		if(reagents.total_volume)
+			. += "<span class='notice'>It has [reagents.total_volume] unit\s left.</span>"
+			return
+		. += "<span class='danger'>It's empty.</span>"
```
Comment (Henri215) [inline code/game/atoms.dm:440]:
```suggestion
		if(reagents.total_volume)
			. += "<span class='notice'>It has [reagents.total_volume] unit\s left.</span>"
		else
			. += "<span class='danger'>It's empty.</span>"
```

File: code/game/atoms.dm [hunk 5]
```diff
@@ -390,24 +390,54 @@
...
+				for(var/I in reagents.reagent_list)
+					var/datum/reagent/R = I
+					. += "<span class='notice'>[R.volume] units of [R] ([round(R.volume / one_percent)]%)</span>"
+				return
+
+		//Otherwise, just show the total volume
+		if(reagents && length(reagents.reagent_list))
+			. += "<span class='notice'>[reagents.total_volume] units of various reagents.</span>"
+			return
+		. += "<span class='notice'>Nothing.</span>"
```
Comment (Henri215) [inline code/game/atoms.dm:433]:
```suggestion
		if(reagents && length(reagents.reagent_list))
			. += "<span class='notice'>[reagents.total_volume] units of various reagents.</span>"
		else
			. += "<span class='notice'>Nothing.</span>"
```

File: code/game/atoms.dm [hunk 6]
```diff
@@ -390,24 +390,54 @@
...
 			else
-				. += "<span class='danger'>It's empty.</span>"
+				blood_type = R.data["blood_type"]
+				. += "<span class='notice'>[R] ([blood_type ? "[blood_type]" : ""]) ([round(R.volume / one_percent)]%)</span>"
+		return
+
+	if(container_type & TRANSPARENT)
+		. += "<span class='notice'>It contains:</span>"
+		if(user.reagent_vision())	// You can see absolute unit quantities of reagents in transparent containers.
+			if(container_type & TRANSPARENT)
```
Comment (Henri215) [inline code/game/atoms.dm:423]: You need to check if its transparent twice here?

File: code/game/atoms.dm (hunk 5 above)
Comment (Oyu07) [reply code/game/atoms.dm:433]: Hijacking this sorry. Can you turn
```

if(reagents && length(...))

```
into
```

if(length(reagent.reagentlist)

```

**Review by FunnyMan3595 — APPROVED** [review]

File: code/game/atoms.dm (hunk 6 above)
Comment (Fordoxia) [reply code/game/atoms.dm]: Oops! Fixed.

File: code/game/atoms.dm (hunk 5 above)
Comment (Fordoxia) [reply code/game/atoms.dm]: > Hijacking this sorry. Can you also turn
> 
> ```
> 
> if(reagents && length(...))
> ```
> 
> into
> 
> ```
> 
> if(length(reagent.reagentlist)
> ```

Done.

File: code/game/atoms.dm [hunk 7]
```diff
@@ -390,24 +390,53 @@
...
+	if(container_type & TRANSPARENT)
+		. += "<span class='notice'>It contains:</span>"
+		if(user.reagent_vision())	// You can see absolute unit quantities of reagents in transparent containers.
+			for(var/I in reagents.reagent_list)
+				var/datum/reagent/R = I
+				. += "<span class='notice'>[R.volume] units of [R] ([round(R.volume / one_percent)]%)</span>"
+			return
+
+		//Otherwise, just show the total volume
+		if(length(reagents.reagent_list))
```
Comment (GDNgit) [inline code/game/atoms.dm:429]:
```suggestion
		if(length(reagents?.reagent_list))
```
null accesses

File: code/modules/clothing/clothing.dm [hunk 8]
```diff
@@ -3,8 +3,12 @@
 	max_integrity = 200
 	integrity_failure = 80
 	resistance_flags = FLAMMABLE
-	var/list/species_restricted = null //Only these species can wear this kit.
-	var/scan_reagents = 0 //Can the wearer see reagents while it's equipped?
+	/// Only these species can wear this kit.
+	var/list/species_restricted = null
```
Comment (GDNgit) [inline code/modules/clothing/clothing.dm:7]:
```suggestion
	var/list/species_restricted
```
don't declare nulls

**Review by Henri215 — APPROVED** [review]

**Review by Burzah — APPROVED** [review]

**Pull request merged**